### PR TITLE
Ren - Fire

### PR DESCRIPTION
**Pull request by RenCarothers**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. |  A method that gathers the top works of a single category and orders them by vote count.
Describe how you approached testing that model method. What edge cases did you come up with? |  I tested a) if a category had more than 10 works, did it display just the top ten?, b) if a category had less than ten, did it display them all, c) if it had no works, did it display a message?
What are `session` and `flash`? What is the difference between them? | Session track the client's browser session (until they close the browser), flash displays a message to the user for one request cycle.  A major difference between the two is that session runs through multiple requests cycles, until the browser is terminated.
What was one thing that you gained more clarity on through this assignment? | Trying to understand different ways you could approach setting up many-to-many, though I would like to practice this more, still.  
What is the Heroku URL of your deployed application? | https://bb-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+class UsersController < ApplicationController
+  def index
+    @users = User.all
+  end
+
+  def show
+    @user = User.find_by(id: params[:id])
+    render_404 unless @user
```
Comment (kaidamasaki) [inline app/controllers/users_controller.rb:8]: Nice use of `render_404`!

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
+class WorksController < ApplicationController
+  before_action :find_work, only: [:show, :edit, :update, :destroy, :upvote]
+  before_action :find_session, only: [:upvote] #TODO: I don't see why I would put this in app. controller if the heroku model only requires login for upvoting? So I put it here
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:3]: The reason you want it in the `ApplicationController` is because you want to display the username at the top of every page if it exists.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      render :edit
+      return
+    end
+  end
+
+  def destroy
+    # find_work
+
+    if @work.nil?
+      redirect_to root_path
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:61]: This code will never run because your filter returns a 404 if the work is nil.

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  CATEGORIES = %w(album book movie) # For checkboxes on form partials // validations
+
+  validates :category,  presence: true,
+            inclusion: { in: CATEGORIES }
+
+  validates :title, presence: true,
+            uniqueness: { scope: :category }
+
+  def self.by_category(category)
+    if Work.all.any? { |work| work.category == category }
```
Comment (kaidamasaki) [inline app/models/work.rb:15]: This is a somewhat inefficient way to find this (you have to pull back all of the works just to check if a category exists) though without writing SQL it would be tricky to optimize.

(It's totally fine here but you will want to avoid patterns like this in production code during your internship.)

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        </li>
+        <li class="nav-item app-header__nav_item">
+          <a class="nav-link" href="/users">View all users</a>
+        </li>
+      </ul>
+      <ul class="nav app-header__user-nav-container">
+
+        <li class="nav-item app-header__nav_item">
+            <% if session[:user_id] %>
+              <% user = User.find(session[:user_id]) %>
```
Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:38]: You should use a controller filter to set an instance variable instead of directly invoking methods on your models.

**Review by kaidamasaki — COMMENTED** [review]
Excellent job!  

There were a few small things you were dinged for but overall your code was clear, clean and DRY.

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️?
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | The homepage didn't use partials to DRY up the column code.
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | No votes fixture.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
